### PR TITLE
Fix node manager failure when ClientTable has a  disconnected entry.

### PR DESCRIPTION
**Pull request by guoyuhong**
## What do these changes do?

ClientTable is Log<ActorID, ActorTableData>, which has multiple entries for a disconnected client. The first one has is_insertion=true and the second one has is_insertion=false. We must make sure this is not a closed client.

When a new raylet starts, the ClientAdded will be called with the disconnected client data. However, since the client was closed, the connection will fail. The following message shows the example log:
```
I0918 17:16:41.168637 2961466240 tables.cc:335] ClientTable::HandleNotification with client: 0e92265aedc9194e66469a0743558f32e77e86f5 is_new=1 is_insertion=1
I0918 17:16:41.170995 2961466240 node_manager.cc:317] [ClientAdded] received callback from client id 0e92265aedc9194e66469a0743558f32e77e86f5
I0918 17:16:41.171082 2961466240 node_manager.cc:328] a new client: 0e92265aedc9194e66469a0743558f32e77e86f5
I0918 17:16:41.171145 2961466240 node_manager.cc:343] [ClientAdded] CONNECTING TO:  30.50.100.140 61838
I0918 17:16:41.171489 2961466240 tables.cc:335] ClientTable::HandleNotification with client: 8fcca9d6b3122effeb2424faa315713b79829ca9 is_new=1 is_insertion=1
I0918 17:16:41.171558 2961466240 node_manager.cc:317] [ClientAdded] received callback from client id 8fcca9d6b3122effeb2424faa315713b79829ca9
I0918 17:16:41.171686 2961466240 node_manager.cc:328] a new client: 8fcca9d6b3122effeb2424faa315713b79829ca9
I0918 17:16:41.171814 2961466240 node_manager.cc:343] [ClientAdded] CONNECTING TO:  30.50.100.140 61881
I0918 17:16:41.172071 2961466240 tables.cc:335] ClientTable::HandleNotification with client: 397783a6eddf33c74007a4cbabb156cbcfd0220d is_new=1 is_insertion=1
I0918 17:16:41.172159 2961466240 node_manager.cc:317] [ClientAdded] received callback from client id 397783a6eddf33c74007a4cbabb156cbcfd0220d
I0918 17:16:41.172209 2961466240 node_manager.cc:328] a new client: 397783a6eddf33c74007a4cbabb156cbcfd0220d
I0918 17:16:41.172261 2961466240 node_manager.cc:343] [ClientAdded] CONNECTING TO:  30.50.100.140 61923
I0918 17:16:41.172420 2961466240 tables.cc:335] ClientTable::HandleNotification with client: 8fcca9d6b3122effeb2424faa315713b79829ca9 is_new=1 is_insertion=0
I0918 17:16:41.172549 2961466240 node_manager.cc:361] [ClientRemoved] received callback from client id 8fcca9d6b3122effeb2424faa315713b79829ca9
I0918 17:16:41.172675 2961466240 tables.cc:335] ClientTable::HandleNotification with client: 397783a6eddf33c74007a4cbabb156cbcfd0220d is_new=1 is_insertion=0
I0918 17:16:41.172762 2961466240 node_manager.cc:361] [ClientRemoved] received callback from client id 397783a6eddf33c74007a4cbabb156cbcfd0220d
I0918 17:16:41.172847 2961466240 tables.cc:335] ClientTable::HandleNotification with client: e3c2435282712dca6d33f715aa7e3a4847a33f50 is_new=1 is_insertion=1
I0918 17:16:41.172936 2961466240 node_manager.cc:317] [ClientAdded] received callback from client id e3c2435282712dca6d33f715aa7e3a4847a33f50
```
Client **`8fcca9d6b3122effeb2424faa315713b79829ca9`** and **`397783a6eddf33c74007a4cbabb156cbcfd0220d`** has 2 entries.
## Related issue number

https://github.com/ray-project/ray/issues/2878

### CODE REVIEW

**Comment by guoyuhong** [comment]
@richardliaw

**Comment by guoyuhong** [comment]
@AmplabJenkins retest this, please.

**Comment by richardliaw** [comment]
So just to make sure I understand, previously we would optimistically update resource tracking before we actually verified that the client was connected? And now we would update the resource tracking only after connection check returns True?

Is it possible to add a test for this?

File: src/ray/raylet/node_manager.cc [hunk 1]
```diff
@@ -334,21 +334,29 @@ void NodeManager::ClientAdded(const ClientTableDataT &client_data) {
...
+    return;
+  }
+
+  // The client is connected.
   auto server_conn = TcpServerConnection(std::move(socket));
   remote_server_connections_.emplace(client_id, std::move(server_conn));
+
+  ResourceSet resources_total(client_data.resources_total_label,
+                              client_data.resources_total_capacity);
+  this->cluster_resource_map_.emplace(client_id, SchedulingResources(resources_total));
```
Comment (robertnishihara) [inline src/ray/raylet/node_manager.cc]: Let's avoid using `this` (I realize this is copied from above).

File: src/ray/raylet/node_manager.cc [hunk 2]
```diff
@@ -334,21 +334,29 @@ void NodeManager::ClientAdded(const ClientTableDataT &client_data) {
...
   boost::asio::ip::tcp::socket socket(io_service_);
-  RAY_CHECK_OK(TcpConnect(socket, client_info.node_manager_address,
-                          client_info.node_manager_port));
+  auto status =
+      TcpConnect(socket, client_info.node_manager_address, client_info.node_manager_port);
+  // ClientTable is Log<ActorID, ActorTableData>, which has multiple entries for
+  // a disconnected client. The first one has is_insertion=true and the second
+  // one has is_insertion=false. We must make sure this is not a close client.
+  if (!status.ok()) {
+    return;
```
[the remark this answers is not in the document]
Comment (robertnishihara) [reply src/ray/raylet/node_manager.cc]: If we return here and don't connect, when will the connection be established?

**Comment by guoyuhong** [comment]
@robertnishihara @richardliaw By my understanding, when a raylet disconnected, it will write a message to ClientTable with `is_insertion=false`. And other raylet client will process it with `ClientRemoved`. But when a new raylet connected in, `ClientAdded` will be called several time with each of the ClientTable entry. For a disconnected client that has two entries in ClientTable, the first call will have `is_insertion=true` and the second call will have `is_insertion=false`. 

Another way to solve this could be cleaning up both entries for the disconnected client.

**Comment by guoyuhong** [comment]
Sorry to mis-express the understanding. For a entry with `is_insertion=false`, it will be processed by `ClientRemoved`. 

The problem is that, after a raylet client disconnected, it only append an entry with `is_insertion=false`, the previous entry with `is_insertion=true` is not removed. Therefore, when a new raylet is connecting, it will receive the first entry which tells the closed client is still alive, and the the connection to the closed client will fail.

File: src/ray/raylet/node_manager.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (guoyuhong) [reply src/ray/raylet/node_manager.cc]: `ClientTableDataT` may come from a disconnected client. The connection should never be established.

**Comment by guoyuhong** [comment]
@robertnishihara Is this PR OK for you? Is `RAY_CHECK(status.message() == "Connection refused")` strange and fragile if the message changes?

**Review by robertnishihara — APPROVED** [review]

**Comment by guoyuhong** [comment]
@robertnishihara Yes. I will add the test in later PR.

**Pull request merged**